Update PS chapter (#340)
r-causal/causal-inference-in-R: chapters/08-propensity-scores.qmd

Below is pseudo-code for using `glm()` to fit a propensity score model using logistic regression.
The first argument is the model, with the exposure on the left side and the confounders on the right.
The `data` argument takes the data frame, and the `family = binomial()` argument denotes that the model should be fit using logistic regression (as opposed to a different generalized linear model, although other links are sometimes used in propensity score).
You've likely fit models like this before, but the key details are that we are predicting the probability of exposure (instead of something around the outcome; we'll get to that!) and that the predictors are the confounders determined from the DAG. 
As we saw in @sec-ci-rct, if we had any predictors of the outcome that weren't related to the exposure, we'd probably want to include those, too, for precision's sake.

```{r}
#| eval: false
glm(
  # predict the probability of treatment
  exposure ~ confounder_1 + confounder_2,
  data = df,
  # using logistic regression
  family = binomial()
)
```

We can extract the propensity scores by pulling out the predictions on the probability scale using `predict()` or `fitted()`. However,

> using the `augment()` function from the [{`broom`}](https://broom.tidymodels.org/) package, we can extract these propensity scores and add them to our original data frame, so we'll use that approach.

Reviewer: ```suggestion
using the `augment()` function from the [{`broom`}](https://broom.tidymodels.org/) package, we can extract these propensity scores and add them to our original data frame in one step, so we'll use that approach.
```

The predictions will be on the *linear logit* scale by default.
Setting the argument `type.predict` to `"response"` indicates that we want to extract the predicted values on the *probability* scale.

> The `data` argument contains the original data frame; if we leave this blank, we'll only get back a data frame with the variables in the propensity score model. However, we need the outcome, too, so it's handy to predict the whole data frame, even though there are no new days in the data set.

Reviewer: ```suggestion
The `data` argument contains the original data frame; if we leave this blank, we'll only get back a data frame with the variables in the propensity score model. However, we need the outcome, too, so it's handy to use the whole data frame, even though there are no new days in the data set.
```

@fig-sim-ps shows simulated scenarios of the kinds of distributions we will tend to see with good, moderate, and poor overlap and balance. Poor balance and overlap can worsen bias and variance and make us less confident in the causal assumptions we need to make.

```{r}
#| label: fig-sim-ps
#| fig-cap: "Simulated distributions of propensity scores. We look for two qualities in mirrored histograms: overlap and balance, related to positivity and exchangeability, respectively."
#| code-fold: true
library(patchwork)

set.seed(2025)

make_model <- function(n = 1000, intercept, slope) {
  df <- tibble(
    z = rnorm(n),
    exposure = rbinom(n, 1, plogis(intercept + slope * z))
  )
  glm(exposure ~ z, data = df, family = binomial)
}

models_overlap <- list(
  good = make_model(intercept = 0, slope = 0.5),
  moderate = make_model(intercept = 0, slope = 1.5),
  poor = make_model(intercept = 0, slope = 3.0)

> )

Reviewer: Love this!!!!!

```{r}
seven_dwarfs_9_with_ps |>
  group_by(park_extra_magic_morning) |>
  reframe(range = range(.fitted))
```

The range between groups seems pretty close. One helpful way to look at the tails is to check the number of *unexposed* observations below the lowest probability for the *exposed* group and the number of *exposed* observations above the highest probability for the *unexposed* group. 

```{r}
seven_dwarfs_9_with_ps |>
  mutate(
    support = case_when(
      park_extra_magic_morning == 0 &
        .fitted < min(.fitted[park_extra_magic_morning == 1]) ~ "unexp_below",
      park_extra_magic_morning == 1 & 
        .fitted > max(.fitted[park_extra_magic_morning == 0]) ~ "exp_above",
      .default = "inside_support",
      .ptype = factor(levels = c("unexp_below", "inside_support", "exp_above"))
    )
  ) |>
  count(support, .drop = FALSE)
```

However, looking at @fig-mirrored-ps, we see some areas of sparsity, implying some combinations have positivity violations. For instance, many more days are without Extra Magic Mornings with a probability below 10%.

```{r}
seven_dwarfs_9_with_ps |>
  count(park_extra_magic_morning, low_prob = .fitted <= .1)
```

> We're likely seeing a combination of structural positivity (some days will never or always get Extra Magic Hours per Disney's decision-making process) and stochastic violations. Since we have a limited number of days in the year and only about 17% of days get Extra Magic Hours in the morning, we expect some sparsity. For instance, @fig-random-days-ps is what the graph would look like if Extra Magic Mornings were randomized with the same proportions of exposed and unexposed days. (If you try this out with different seeds, you'll see that these data are prone to such random violations.) Propensity score methods are more susceptible to issues with positivity violations than outcome model-based methods and doubly robust methods; we'll have to keep an eye on this.

Reviewer: Is this DR part of this true? I thought in general DR will inherit the huge variance caused by a positivity violation from the PS model but I could be wrong! Definitely true that parametric outcome models are not as susceptible though

Author: It is true for TMLE (one of its many benefits!), but you're right that it's not for AIPW. Not sure about DoubleML. I'll clarify

We can calculate many different weights depending on the target estimand of interest (see @sec-estimands for details).
In this section, we will focus on the Average Treatment Effect (ATE) weights, commonly referred to as inverse probability weights.
The weight is constructed as follows: each observation is weighted by the *inverse* of the probability of receiving the exposure it actually received.

$$w_{ATE} = \frac{X}{p} + \frac{(1 - X)}{1 - p}$$

For example, if observation 1 had a very high likelihood of being exposed given their pre-exposure covariates ($p = 0.9$), but they in fact were *not* exposed, their weight would be 10 ($w_1 = 1 / (1 - 0.9)$).
Likewise, if observation 2 had a very high likelihood of being exposed given their pre-exposure covariates ($p = 0.9$), and they *were* exposed, their weight would be 1.1 ($w_2 = 1 / 0.9$).
Intuitively, we give more weight to observations that, based on their measured confounders, appear to have helpful information for constructing a counterfactual---we would have predicted that they were exposed, but by chance, they were not, or vice versa. 

> The `{propensity}` package calculates a variety of propensity score weights that follow the pattern `wt_estimand()`. To calculate the ATE, we use `wt_ate()`, which we provide the fitted propensity score and the observed exposure values. 

Reviewer: ```suggestion
The `{propensity}` package calculates a variety of propensity score weights with functions named to follow the pattern `wt_estimand()`. To calculate the ATE, we use `wt_ate()`, which we provide the fitted propensity score and the observed exposure values. 
```

```{r}
#| label: tbl-df-wt
#| tbl-cap: >
#|   The first six observations in the `seven_dwarfs_9_with_wt` dataset, including their propensity scores in the `.fitted` column and weight in the `w_ate` column.
#| code-fold: true
seven_dwarfs_9_with_wt |>
  select(
    park_date,
    park_extra_magic_morning,
    park_ticket_season,
    park_close,
    park_temperature_high,
    .fitted,
    w_ate
  ) |>
  head() |>
  gt() |>
  cols_label(
    .fitted = "P(EMM)",
    park_date = "Date",
    park_extra_magic_morning = "Extra Magic Mornings",
    park_ticket_season = "Ticket Season",
    park_close = "Close time",
    park_temperature_high = "Historic Temp.",
    w_ate = "ATE Weights"
  )
```

::: {.callout-note}
## WeightIt

> If you like the feel of MatchIt, it has a cousin package called `{WeightIt}` with the same design principles and many useful features. We'll focus on propensity, but WeightIt is easy to use if you are familiar with MatchIt.

Reviewer: Yay! Love this thank you for adding it in!
